### PR TITLE
linter: added a new way of return type inference for functions and methods

### PR DESCRIPTION
**Pull request by i582**
The types are now inferred as follows:
1. If there is an `@param` annotation, then its value becomes
   the return type;

2. If there is a type hint, then it is added to the types
   from the `@param`. If the `@param` is empty, then the type
   matches the type hint itself;

3. If there is no `@param` annotation and type hint, then
   the return type is equal to the union of the types that
   are returned from the function by return.

Part of #849 

The next step, adding type checking.

### CODE REVIEW

File: src/linter/utils.go [hunk 1]
```diff
@@ -90,6 +90,32 @@ func typesMapToTypeExpr(p *phpdoc.TypeParser, m meta.TypesMap) phpdoc.Type {
 	return p.Parse(typeString)
 }
 
+// functionReturnType returns the return type of a function over computed types
+// according to the convention below:
+//
+// The types are inferred as follows:
+// 1. If there is an @param annotation, then its value becomes the return type;
```
Comment (quasilyte) [inline src/linter/utils.go:97]: Maybe `@return` annotation was implied here (and below)?

File: src/linter/utils.go [hunk 2]
```diff
@@ -90,6 +90,32 @@ func typesMapToTypeExpr(p *phpdoc.TypeParser, m meta.TypesMap) phpdoc.Type {
...
+//
+// 2. If there is a type hint, then it is added to the types from the @param.
+//    If the @param is empty, then the type matches the type hint itself;
```
Comment (quasilyte) [inline src/linter/utils.go:100]: Does merging type hint and `@return` improve the type inference?

For example, in new tests there is `f6()` which has `array` type hint and `int[]` from phpdoc annotation.
I would guess that marking `f6()` result as `int[]` is more desirable than `int[]|mixed[]`.

File: src/linter/utils.go (hunk 1 above)
Comment (i582) [reply src/linter/utils.go]: Yep. Done.

**Comment by quasilyte** [comment]
1. Please increase the cache version as we changed the way how function results are inferred.
2. Add something in a cache_test so we get a different result for the type inference before and after this change.

I believe the cache_test is not failing because we don't have a function there that would be inferred differently.

File: src/linter/utils.go (hunk 2 above)
Comment (i582) [reply src/linter/utils.go]: Done

**Review by quasilyte — APPROVED** [review]

**Pull request merged**